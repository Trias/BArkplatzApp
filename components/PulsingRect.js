import React from 'react';
import {Animated } from 'react-native';
import { Rect } from 'react-native-svg';

function getInitialState() {
    const anim = new Animated.Value(0.3);
    const fillOpacity = anim.interpolate({
        inputRange: [0, 1],
        outputRange: [0, 1],
    });

    return { anim, fillOpacity };
}

const AnimatedRect = Animated.createAnimatedComponent(Rect);

export default class PulsingRect extends React.Component {
    state = getInitialState();

    componentDidMount() {
        this.startAnimation();
    }

    startAnimation() {
        const { anim } = this.state;
        Animated.timing(anim, {
            toValue: 0.3,
            duration: 1000,
            useNativeDriver: true,
        }).start(() => {
            this.secondAnimation()
        });
    }

    secondAnimation() {
        const { anim } = this.state;
        Animated.timing(anim, {
            toValue: 0.8,
            duration: 1000,
            useNativeDriver: true,
        }).start(() => {
            this.startAnimation()
        });
    }

    render() {
        let { x, y, width, height, fill } = this.props;
        const { fillOpacity } = this.state;
        return <AnimatedRect
                fillOpacity={fillOpacity}
                x={x}
                y={y}
                width={width}
                height={height}
                fill={fill}
            />
    }
}
