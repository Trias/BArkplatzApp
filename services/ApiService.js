const ACCESS_TOKEN = '';
const URL = ``;

import dummyData from '../data/UpdatedDummy.json';

let cache = null;

let interval = null;

async function rpcCall(method, params = {}) {
    let response;
    try {
        console.log('start API call: '+ method + ' with params: ' + JSON.stringify(params));
        response = await fetch(URL, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                method, params
            })
        });
        let json = await response.json()

        console.log('api call successful');
        console.log('method: ' + method);
        console.log('params: ' + JSON.stringify(params));
        console.log(json);

        return json;
    } catch (e) {
        console.warn('error while getting API data')
        console.log(response);
        console.log(e);
    }
}

export async function initApi() {
    await fetchAllAsync();
    interval = setInterval(fetchAllAsync, 60 * 1000);
}
export function stopApi() {
    clearInterval(interval);
    interval = null;
}
export async function fetchAllAsync() {
    let response = await rpcCall('fetchAll');
    if (response && !response.error) {
        cache = addCoordinates(response)
    } else {
        console.warn('empty data from api, using old data');
    }
    return cache;
}
export async function getReservationsOfThisUserAsync() {
    return await rpcCall('getReservationsOfThisUser')
}
export async function reserveAsync(reservation) {
    return await rpcCall('reserve', reservation)
}
export async function deleteReservationAsync(reservation) {
    return await rpcCall('deleteReservation', reservation);
}
export async function getReservationOfOneParkingSpaceAsync(parkingSpaceId) {
    return await rpcCall('getReservationOfOneParkingSpace', { parkingSpaceId });
}
export function getData() {
    if (!cache) {
        console.warn('error fetching data from api, using dummy data');
        cache = addCoordinates(dummyData);
    }

    return cache;
}

function addCoordinates(cache){
    for (let parkingLot of Object.values(cache)) {
        let firstParkingSpace = Object.values(parkingLot.spaces)[0];
        parkingLot.latitude = firstParkingSpace.latitude;
        parkingLot.longitude = firstParkingSpace.longitude;
    }

    return cache
}
