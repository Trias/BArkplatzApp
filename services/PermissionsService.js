import * as Permissions from 'expo-permissions';

let hasLocationPermission = false;
export async function requestLocationPermission() {
    if (hasLocationPermission) {
        return true;
    }
    try {
        const { status } = await Permissions.askAsync(Permissions.LOCATION);
        if (status === 'granted') {
            hasLocationPermission = true;
            console.log('gps granted');
            return true;
        } else {
            console.log('gps permission denied');
            return false;
        }
    } catch (err) {
        console.warn(err);
        return false;
    }
}

let hasNotificationPermission = false;
export async function requestNotificationPermission() {
    if (hasNotificationPermission) {
        return true;
    }
    try {
        const { status } = await Permissions.askAsync(Permissions.USER_FACING_NOTIFICATIONS);
        if (status === 'granted') {
            hasNotificationPermission = true;
            console.log('notifications granted');
            return true;
        } else {
            console.log('notifications denied');
            return false;
        }
    } catch (err) {
        console.warn(err);
        return false;
    }
}