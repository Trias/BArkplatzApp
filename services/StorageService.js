import { AsyncStorage } from "react-native";

const cache = {};

export async function initStorage(){
    const allKeys = await AsyncStorage.getAllKeys();
    for (let key of allKeys){
        let value = await AsyncStorage.getItem(key);
        cache[key] = value;
    }
}

export function getItem(name){
    return cache[name];
}

export function setItem(name, item){
    cache[name] = item;
    setItemAsync(name, item);
}

export function removeItem(name){
    delete cache[name];
    removeItemAsync(name);
}

export async function getItemAsync(name){
    return AsyncStorage.getItem(name)
}

export async function setItemAsync(name, item){
    cache[name] = item;
    return AsyncStorage.setItem(name, item);
}

export async function removeItemAsync(name){
    cache[name] = undefined;
    return AsyncStorage.removeItem(name);
}