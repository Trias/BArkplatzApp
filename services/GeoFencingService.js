import * as TaskManager from 'expo-task-manager';
import * as Location from 'expo-location';
import { Notifications } from 'expo';
import { requestLocationPermission, requestNotificationPermission } from './PermissionsService';

const TASK_NAME = "parkingLots";

let geoFencingActive = false;
export async function initGeofencing() {
  try {
    // permissions returns only for location permissions on iOS and under certain conditions, see Permissions.LOCATION
    let locationStatus = await requestLocationPermission();
    let notificationStatus = await requestNotificationPermission();

    if (!geoFencingActive) {
      if (locationStatus && notificationStatus) {
        geoFencingActive = true;
        await Location.startGeofencingAsync(TASK_NAME, [{
          // erba: 49.903, 10.87
          // harcoded for now to avoid dependency on network request
          identifier: "631", // heinrichsdamm
          latitude: 49.89,
          longitude: 10.903,
          radius: 100
        },
        {
          identifier: "38", // brennerstraÃŸe
          latitude: 49.901,
          longitude: 10.90,
          radius: 100
        }]);
      } else {
        console.log('Location permission or notification permission not granted');
      }
    }
  } catch (e) {
    console.log('error while geofencing');
  }
}

TaskManager.defineTask(TASK_NAME, ({ data: { eventType, region }, error }) => {
  console.log("geofencing event");
  if (error) {
    // check `error.message` for more details.
    console.log(error)
    return;
  }
  if (eventType === Location.GeofencingEventType.Enter) {
    const localNotification = {
      title: "Sie sind am Parkplatz angekommen!",
      body: "jetzt Parkplatz merken"
    }
    Notifications.addListener(notification => {
      if (notification.origin == "selected") {
        let { MainTabNavigator } = import('../navigation/MainTabNavigator');
        MainTabNavigator.navigation.navigate('ParkingLotDetail', { parkingLotId: region.identifier })
      }
    })
    Notifications.presentLocalNotificationAsync(localNotification);
    console.log("You've entered region:", region);
  } else if (eventType === Location.GeofencingEventType.Exit) {
    console.log("You've left region:", region);
  }
});