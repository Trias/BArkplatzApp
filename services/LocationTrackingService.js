import * as Location from 'expo-location';
import {requestLocationPermission} from './PermissionsService';

let locationTrackingInterval = null;
let locationTrackingActive = false;

export async function initLocationTracking(){
  try {
    // permissions returns only for location permissions on iOS and under certain conditions, see Permissions.LOCATION
    let locationStatus = await requestLocationPermission();
    if (!locationTrackingActive && locationStatus){
      locationTrackingActive = true;
      locationCache = await getLocationAsync();
      locationTrackingInterval = setInterval(getLocationAsync, 10000);
    }
  }catch(e){
    console.log('error while location');
  }
}

export function stopLocationTracking(){
  clearInterval(locationTrackingInterval);
  locationTrackingInterval = null;
  locationTrackingActive = false;
}

let locationCache = null;
export async function getLocationAsync(){
  locationCache = await Location.getCurrentPositionAsync({enableHighAccuracy: true});
  return locationCache;
}

export function getLocation(){
  if(locationCache){
    return locationCache;
  }else{
    console.warn('could not get location data, using fallback');
    return {coords: {latitude: 49.9028939, longitude: 10.8686975}};
  }
}