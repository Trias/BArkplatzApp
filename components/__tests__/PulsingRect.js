import React from 'react';
import renderer from 'react-test-renderer';

import PulsingRect from '../PulsingRect';
jest.useFakeTimers();

import * as ApiService from '../../services/ApiService';
import * as LocationTrackingService from '../../services/LocationTrackingService';

jest.mock('../../services/ApiService');
jest.mock('../../services/LocationTrackingService');

function addCoordinates(cache){
    for (let parkingLot of Object.values(cache)) {
        let firstParkingSpace = Object.values(parkingLot.spaces)[0];
        parkingLot.latitude = firstParkingSpace.latitude;
        parkingLot.longitude = firstParkingSpace.longitude;
    }

    return cache
}
ApiService.getData = jest.fn(() => {return addCoordinates(require('../../data/UpdatedDummy'))});


it(`renders correctly`, () => {
  const tree = renderer.create(<PulsingRect
  />).toJSON();

  expect(tree).toMatchSnapshot();
});
