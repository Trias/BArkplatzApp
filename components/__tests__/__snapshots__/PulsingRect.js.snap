// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RNSVGRect
  fill={
    Array [
      0,
      4278190080,
    ]
  }
  fillOpacity={0.3}
  fillRule={1}
  height={0}
  matrix={
    Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ]
  }
  opacity={1}
  originX={0}
  originY={0}
  propList={
    Array [
      "fillOpacity",
    ]
  }
  rotation={0}
  rx={0}
  ry={0}
  scaleX={1}
  scaleY={1}
  skewX={0}
  skewY={0}
  stroke={null}
  strokeDasharray={null}
  strokeDashoffset={null}
  strokeLinecap={0}
  strokeLinejoin={0}
  strokeMiterlimit={4}
  strokeOpacity={1}
  strokeWidth={1}
  vectorEffect={0}
  width={0}
  x={0}
  y={0}
/>
`;
